ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/views/HiveRowIsDeletedPropagator.java

```diff
@@ -92,86 +159,155 @@ public RelNode visit(HiveTableScan scan) {
+  public RelNode visit(HiveJoin join, Context context) {
+    // Propagate columns to left input
+    RelNode tmpJoin = visitChild(join, 0, join.getInput(0), context);
+    RelNode newLeftInput = tmpJoin.getInput(0);
+    RelDataType newLeftRowType = newLeftInput.getRowType();
+    // Propagate columns to right input.
+    // All column references should be shifted in candidate predicates to the left
+    Context rightContext = new Context();
+    int originalLeftFieldCount = join.getInput(0).getRowType().getFieldCount();
+    for (Map.Entry<Integer, RexNode> entry : context.rowIdPredicates.entrySet()) {
+      if (entry.getKey() > originalLeftFieldCount) {
+        rightContext.rowIdPredicates.put(entry.getKey() - originalLeftFieldCount,
+          new InputRefShifter(originalLeftFieldCount, -originalLeftFieldCount, relBuilder).apply(entry.getValue()));
+      }
+    }
+    tmpJoin = visitChild(join, 1, join.getInput(1), rightContext);
+    RelNode newRightInput = tmpJoin.getInput(1);
+    RelDataType newRightRowType = newRightInput.getRowType();
+
+    // Create input refs to propagated columns in left and right inputs
+    int rightAnyDeletedIndex = newRightRowType.getFieldCount() - 2;
+    int rightAnyInsertedIndex = newRightRowType.getFieldCount() - 1;
     RexBuilder rexBuilder = relBuilder.getRexBuilder();
-    RexNode leftRowIsDeleted = rexBuilder.makeInputRef(
-            leftRowType.getFieldList().get(leftRowIsDeletedIndex).getType(), leftRowIsDeletedIndex);
-    RexNode rightRowIsDeleted = rexBuilder.makeInputRef(
-            rightRowType.getFieldList().get(rightRowIsDeletedIndex).getType(),
-            leftRowType.getFieldCount() + rightRowIsDeletedIndex);
+    RexNode leftDeleted = createInputRef(newLeftInput, 2);
+    RexNode leftInserted = createInputRef(newLeftInput, 1);
+    RexNode rightDeleted = rexBuilder.makeInputRef(
+        newRightRowType.getFieldList().get(rightAnyDeletedIndex).getType(),
+        newLeftRowType.getFieldCount() + rightAnyDeletedIndex);
+    RexNode rightInserted = rexBuilder.makeInputRef(
+        newRightRowType.getFieldList().get(rightAnyInsertedIndex).getType(),
+        newLeftRowType.getFieldCount() + rightAnyInsertedIndex);
 
+    // Shift column references refers columns coming from right input in join condition since the new left input
+    // has a new columns
     RexNode newJoinCondition;
     int newLeftFieldCount;
-    if (join.getInput(0).getRowType().getField(VirtualColumn.ROWISDELETED.getName(), false, false) == null) {
-      // Shift column references refers columns coming from right input by one in join condition since the new left input
-      // has a new column
-      newJoinCondition = new InputRefShifter(leftRowType.getFieldCount() - 1, relBuilder)
-          .apply(join.getCondition());
-
-      newLeftFieldCount = leftRowType.getFieldCount() - 1;
+    int shift;
+    if (newLeftRowType.getField("_rowId_greater_than", false, false) == null) {
```

Comment:
is there a constant for "_rowId_greater_than"

Reply:
Removed this if statement because only the true branch is executed.